### ISSUE TITLE
Incorrectly blocks exporting with ".tsv" or ".cfg" extension

### ISSUE THREAD

**Issue by alexchandel** [sqlitebrowser/sqlitebrowser#1880]
Details for the issue
--------------------

#### What did you do?

Attempting to export data using "Export data as CSV" to a file with an extension other than ".csv" fails.  This is especially nonsensical when exporting with tab delimiters, i.e. a `.tsv` file.  Every file I export requires a name other than `.csv`, and it's tedious to work around DB4S's blocking me.

#### What did you expect to see?

Exporting "as CSV" should allow any file extension.  No one, especially no one using a sqlite browser, needs a program forcing them which file extension to use.

#### What did you see instead?

DB Browser blocks this.

Useful extra information
-------------------------

The info below often helps, please fill it out if you're able to. :)

#### What operating system are you using?

- [x] Mac OS: ( _version: 10.14.4 )

#### What is your DB4S version?

- [x] 3.11.2

#### Did you also

- [x] Try out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [x] Search for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by scottfurry** [comment]
This should be do-able. Tracing down the call, an edit to `ExportDataDialog::accept()` (case stmt @ line 303 `[prj root]/src/ExportDataDialog.cpp) looks like all that is needed (someone from the studio audience correct me if I'm wrong).

The default will still be ".csv", but the dialog file filter would be extended to enable/accept the requested file extension.

**Comment by justinclift** [comment]
This sounds reasonable too. :smile:

**Comment by scottfurry** [comment]
I can find references for `.tsv` (tab-separated value) and `.dsv` (delimiter-separated value).
I can't find references for `.cfg` file extension (other than configuration files - not CSV domain).

**Comment by justinclift** [comment]
Looking at those lines, it does seem like the right area: https://github.com/sqlitebrowser/sqlitebrowser/blob/cec6b825611a5aa44a7fe09e22ccaca20ecb5510/src/ExportDataDialog.cpp#L303-L306

Not sure why it's blocking any manually typed in extension though.  We've previously seen some weirdness on just macOS, where the extension filters to show matching files wouldn't work.

But that's not this, and it shouldn't be happening these days anyway.  In theory. :wink:

**Comment by alexchandel** [comment]
@scottfurry It's not common, but there's no reason why it shouldn't be allowed.  There's no reason why `.idowhatiwant` shouldn't be allowed.  A text file's extension is a user decision (a user competent enough to know what SQL is), not something to be mandated.

To be sure, there shouldn't be an additional confirmation for alternate endings at all.  macOS explicitly highlights the file's basename in the save dialog.  If you (someone competent enough to know what SQL is) go out of your way to change the extension, it should silently accept it.  This is the standard behavior for macOS save dialogs.

For reference, this is the prompt:
![Screen Shot](https://user-images.githubusercontent.com/637714/57385330-e34c8300-7177-11e9-9a75-3144e1858d4e.png)

**Comment by justinclift** [comment]
Oh.  That looks like macOS itself throwing up a warning, not something from us.

Wonder what's causing that then...

**Comment by justinclift** [comment]
Doing some searching online, there are lots of reports of people having trouble saving files with (various) extensions on macOS.

It **seems** to be something to do with macOS understanding what a CSV file is, and feeling very strongly about how they should be named.

Not sure if there's anything we can do about it. :frowning:

**Comment by alexchandel** [comment]
That isn't the case, nor is that how Cocoa works.

For reference, other macOS apps don't have a problem with nonstandard endings; TextEdit has a slightly more sophisticated prompt, but happily saves this file:  ![Screen Shot](https://user-images.githubusercontent.com/637714/57388770-87d1c380-717e-11e9-8624-432d1dba6377.png)
Atom does the same, but it has no file-extension options at all in its save prompt.

**Comment by alexchandel** [comment]
Moreover, this is also impossible in Windows.  When you try to export a table to CSV with `.idowhatiwant` as your ending, it saves a file called `NAME.idowhatiwant.csv`.  You're telling the save prompt that you don't permit arbitrary endings.

**Comment by justinclift** [comment]
Hmmm.  The text string "You cannot save this document with" (etc) doesn't appear in our source code anywhere.  That exact string is turning up in searches online for other application too, also on macOS.

So, I'm definitely confused now. :wink:

**Comment by justinclift** [comment]
There is mention around of some association between allowed extensions, and the application plist file.

But we don't have `csv` in [ours](https://github.com/sqlitebrowser/sqlitebrowser/blob/master/src/app.plist), so that doesn't seem to match either.

Ideas are very much welcome. :smile:

**Comment by justinclift** [comment]
Just tried it on windows, and yep, you're right.  We're definite appending .csv to the end of the filename string there.  That does seem a bit weird. :frowning:

**Comment by scottfurry** [comment]
@justinclift  - I'm in that bit of code now.
[prj src]/src/ExportEditDialog.cpp - line 306;line 310. Assignment to variable `default_file_extension`. Variable is then passed on to a QFileDialog that uses contents to append as default to filename user requests. I believe this can be overwritten during the QFileDialog operation if the user wants to add a different extension, but I could be wrong here.

tl;dr - Actions of the QFileDialog, which has to contend with how EACH PLATFORM(MacOS,Linux,WIN) does file save/open dialogs.

**Comment by justinclift** [comment]
Ahhh yeah.  Found it.  Looks like an unconditional append of the extension:

https://github.com/sqlitebrowser/sqlitebrowser/blob/cec6b825611a5aa44a7fe09e22ccaca20ecb5510/src/ExportDataDialog.cpp#L348

Also a few lines down:

https://github.com/sqlitebrowser/sqlitebrowser/blob/cec6b825611a5aa44a7fe09e22ccaca20ecb5510/src/ExportDataDialog.cpp#L371

@scottfurry You want to fix it?  I'm about to hit the sack, so can't really motivate myself to start the dev environment (etc) atm. :wink:

**Comment by scottfurry** [comment]
Then there is this in [API docs](https://doc.qt.io/qt-5/qfiledialog.html#getSaveFileName):

> On Windows, and macOS, this static function will use the native file dialog and not a QFileDialog.

**Comment by scottfurry** [comment]
I can tweak filters. That's easy enough. But overriding OS behaviour seems like a non-starter.
I'll do what I can but no promises.

**Comment by justinclift** [comment]
Wouldn't just removing the `+ default_file_extension` at the end of those fix the problem?

**Comment by scottfurry** [comment]
Not sure. We would have to test on each platform to see what happens.
I'll leave the variable in *for now*. We can then find volunteers, and/or human guinea pigs, to test what removing the variable does for each platform.

**Comment by justinclift** [comment]
That's what the nightly builds are for. :smile:

**Comment by justinclift** [comment]
On that note, I need to get the macOS nightly builds working again soon. ish. :wink:

**Comment by scottfurry** [comment]
Good Luck. I only have access to a Late 2010 that can't be upgraded to Mojave and my test environment is not up to date.

**Comment by justinclift** [comment]
Heh Heh Heh.  That's pretty much what our build mac is:

https://everymac.com/systems/apple/mac_mini/specs/mac-mini-core-2-duo-2.4-mid-2010-specs.html

I know one way to fix the problem, I'm just expecting it'll take most of a day to do.  And other stuff is higher priority atm. :smile:

**Comment by justinclift** [comment]
@scottfurry If you ever need to get Mojave running on that system, it'll run fine in a VM on the system.  Just not on the bare hardware itself. :wink:

**Comment by alexchandel** [comment]
> Hmmm. The text string "You cannot save this document with" (etc) doesn't appear in our source code anywhere. That exact string is turning up in searches online for other application too, also on macOS.

That's because Aqua requires predefined, pretranslated strings for most builtin responses and errors, to present a uniform user experience.  Cocoa uses this one **because you are telling it to forbid other file extensions**.  That's how Cocoa works.  That's not because of a plist or anything else confused devs claim in forums.  The relevant option is [`NSSavePanel.allowsOtherFileTypes: Bool`](https://developer.apple.com/documentation/appkit/nssavepanel/1526960-allowsotherfiletypes).  Stop blaming macOS for doing what you told it to do.

You'll have to check how `QFileDialog`'s source sets or clears this, but I'd hazard that however Qt is configured to let "*.*" work on windows will also work for macOS.

**Comment by scottfurry** [comment]
@alexchandel,
From earlier today...
> Then there is this in [API docs](https://doc.qt.io/qt-5/qfiledialog.html#getSaveFileName):
> 
> > On Windows, and macOS, this static function will use the native file dialog and not a QFileDialog.

There is more than Mac to contend with here. This is not a simple or easy manipulation as multiple platforms are of concern. We are fortunate that we can leverage Qt to handle most of the issues in cross-platform usage. However, sometimes we are limited by the libraries we have available.

We will do the best that we can with what we have in front of us.

**Comment by alexchandel** [comment]
@scottfurry Right but I'd hazard that however it can be configured to permit `*.*` on Windows may work with macOS, even if it claims to ignore the filter, judging from [this bug](https://bugreports.qt.io/browse/QTBUG-44227?attachmentOrder=desc)

**Comment by mgrojo** [comment]
> Wouldn't just removing the `+ default_file_extension` at the end of those fix the problem?

That's only the default filename provided to the user in the dialog. In the case of Linux, you can later change the extension to whatever you want.

Maybe the solution for other platforms is adding the typical filters, something like this:
"Comma separated values (*.csv);;Tab separated values (*.tsv);;Delimiter separated values (*.dsv);;Text files (*.txt);;All files (*)"

In Linux, it only has effect on the displayed files in the dialog, the filename used by the user is not modified in any way.

If it's the same in other platforms, we could get the selected filter (parameter  `QString *selectedFilter = nullptr`) and add/fix the extension before saving the file. But maybe only by allowing more filters, makes the OS to add the extension by itself based on the selected filter.

**Comment by chrisjlocke** [comment]
> Moreover, this is also impossible in Windows.

I'm late to the party, but thought I'd join the fun.
If you type a filename with quotes, Windows will save as that file, and not append the default extension.  So if you typed `filename`, `.csv` will get appended.  If you typed `"filename"`, then it will save as `filename.` (with no extension)

**Comment by scottfurry** [comment]
@chrisjlocke - shows my age. I forgot about this little *trick* in MSW.

**Comment by mgrojo** [comment]
@scottfurry Isn't this ready to be closed after merging  #1885? @alexchandel could you confirm using a [nightly build](https://nightlies.sqlitebrowser.org/latest/)?

**Comment by scottfurry** [comment]
@mgrojo Yup. I had though change of source code addressed this issue. If original issue is addressed, I think we can close.

**Comment by MKleusberg** [comment]
Cool, let's close this then :smile:
